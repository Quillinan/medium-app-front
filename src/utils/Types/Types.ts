export interface AuthContextType {
  setAuth: (authToken: string | null) => void;
  removeAuth: () => void;
  token: string | null;
}

export interface ErrorResponse {
  code: string;
  message: string;
  detailedMessage: string;
  helpUrl: string;
  details: null;
}

export interface Birthday {
  NOME: string;
  DTNASCIMENTO: number;
}

export interface TableBirthdayProps {
  data: { NOME: string; DTNASCIMENTO: number }[];
}

export interface Post {
  id: number;
  title: string;
  subtitle: string;
  content: string;
  createdAt: Date;
  updatedAt: Date;
  authorId: string;
  authorName: string;
}
