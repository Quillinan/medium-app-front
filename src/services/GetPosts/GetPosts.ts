import handleError from '@errors/HandleError/HandleError';
import { apiUrl, get } from '@services/Api/Api';
import { Post, ErrorResponse } from '@utils/Types/Types';

export const getPosts = async (): Promise<
  Post[] | ErrorResponse | undefined
> => {
  const url = `${apiUrl}api/Post`;
  try {
    const response = await get(url);
    return response as Post[];
  } catch (error: any) {
    return handleError(error);
  }
};
