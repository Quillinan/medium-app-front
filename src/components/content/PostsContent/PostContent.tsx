import React, { useEffect, useState } from 'react';
import { Post, ErrorResponse } from '@utils/Types/Types';
import { getPosts } from '@services/GetPosts/GetPosts';

const PostsContent: React.FC = () => {
  const [posts, setPosts] = useState<Post[]>([]);
  const [error, setError] = useState<ErrorResponse | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      const data = await getPosts();

      if ((data as ErrorResponse).code) {
        setError(data as ErrorResponse);
      } else {
        setPosts(data as Post[]);
      }
    };

    fetchData();
  }, []);

  if (error) {
    return <div>Erro ao carregar posts: {error.message}</div>;
  }

  return (
    <div data-testid='posts-content' className='space-y-8'>
      {posts.map(post => (
        <div
          key={post.id}
          className='flex flex-col border-b border-gray-200 pb-8'
        >
          <h2 className='mt-1 text-2xl font-semibold text-gray-900'>
            {post.title}
          </h2>
          <p className='mt-2 text-gray-700'>{post.content.slice(0, 100)}...</p>
          <p className='mt-1 text-sm text-gray-500'>
            Criado em: {new Date(post.createdAt).toLocaleDateString()}
          </p>
          <p className='mt-1 text-sm text-gray-500'>
            Autor ID: {post.authorId}
          </p>
        </div>
      ))}
    </div>
  );
};

export default PostsContent;
