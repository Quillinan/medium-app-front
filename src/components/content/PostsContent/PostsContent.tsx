import React, { useEffect, useState, useCallback } from 'react';
import { Post, ErrorResponse } from '@utils/Types/Types';
import { getPosts } from '@services/GetPosts/GetPosts';
import PostContentList from '@components/PostContentList/PostContent';
import { showLoading } from '@utils/LoadingHelper/LoadingHelper';

const PostsContent: React.FC = () => {
  const [data, setData] = useState<Post[]>([]);
  const [error, setError] = useState<string | null>(null);

  const fetchData = useCallback(async () => {
    try {
      const response = await showLoading(getPosts());

      if ((response as ErrorResponse).code) {
        setError((response as ErrorResponse).message || 'Erro desconhecido');
      } else {
        setData(response as Post[]);
      }
    } catch (error) {
      setError('Erro ao carregar os dados da API');
      console.error(error);
    }
  }, []);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return (
    <div
      data-testid='posts-content'
      className='flex items-center justify-center h-full mt-4'
    >
      {error ? <p>{error}</p> : <PostContentList data={data} />}
    </div>
  );
};

export default PostsContent;
