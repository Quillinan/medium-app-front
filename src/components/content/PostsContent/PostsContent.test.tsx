import { act, render, screen, waitFor } from '@testing-library/react';
import { Post } from '@utils/Types/Types';
import { getPosts } from '@services/GetPosts/GetPosts';
import { showLoading } from '@utils/LoadingHelper/LoadingHelper';
import { Mock } from 'vitest';
import PostsContent from './PostsContent';

vi.mock('@services/GetPosts/GetPosts', () => ({
  getPosts: vi.fn(),
}));

vi.mock('@utils/LoadingHelper/LoadingHelper', () => ({
  showLoading: vi.fn(),
}));

describe('PostsContent', () => {
  const mockOnPostSelect = vi.fn();

  const mockPosts: Post[] = [
    {
      id: 1,
      title: 'First Post',
      subtitle: 'Subtitle of First Post',
      content: 'This is the content of the first post.',
      createdAt: new Date('2024-10-28T00:00:00Z'),
      updatedAt: new Date('2024-10-28T00:00:00Z'),
      authorId: '123',
      authorName: 'Author One',
      coverImageUrl: 'http://example.com/image1.jpg',
    },
    {
      id: 2,
      title: 'Second Post',
      subtitle: 'Subtitle of Second Post',
      content: 'This is the content of the second post.',
      createdAt: new Date('2024-10-29T00:00:00Z'),
      updatedAt: new Date('2024-10-29T00:00:00Z'),
      authorId: '456',
      authorName: 'Author Two',
      coverImageUrl: 'http://example.com/image2.jpg',
    },
  ];

  it('should display a generic error message if fetching data fails', async () => {
    const errorMessage = 'Erro ao carregar os dados da API';
    (getPosts as Mock).mockRejectedValue(new Error('Erro no servidor'));

    render(<PostsContent onPostSelect={mockOnPostSelect} />);

    expect(showLoading).toHaveBeenCalled();
    const error = await screen.findByText(errorMessage);
    expect(error).toBeInTheDocument();
  });

  it('should display an error message if fetching data fails', async () => {
    const errorMessage = 'Erro ao carregar os dados da API';
    (getPosts as Mock).mockRejectedValue(new Error(errorMessage));
    (showLoading as Mock).mockRejectedValue(new Error(errorMessage));

    const consoleErrorSpy = vi
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(<PostsContent onPostSelect={mockOnPostSelect} />);

    expect(showLoading).toHaveBeenCalled();
    const error = await screen.findByText(errorMessage);
    expect(error).toBeInTheDocument();

    consoleErrorSpy.mockRestore();
  });

  it('should display post data when fetching is successful', async () => {
    (getPosts as Mock).mockResolvedValue(mockPosts);
    (showLoading as Mock).mockImplementation(promise => promise);

    await act(async () => {
      render(<PostsContent onPostSelect={mockOnPostSelect} />);
    });

    expect(showLoading).toHaveBeenCalled();

    await waitFor(() => {
      expect(screen.getByText('First Post')).toBeInTheDocument();
      expect(screen.getByText('Subtitle of First Post')).toBeInTheDocument();
      expect(screen.getByText('Autor : Author One')).toBeInTheDocument();
      expect(screen.getByText('Second Post')).toBeInTheDocument();
      expect(screen.getByText('Subtitle of Second Post')).toBeInTheDocument();
      expect(screen.getByText('Autor : Author Two')).toBeInTheDocument();
    });
  });

  it('should display posts data sorted by date with the most recent first', async () => {
    (getPosts as Mock).mockResolvedValue(mockPosts);
    (showLoading as Mock).mockImplementation(promise => promise);

    await act(async () => {
      render(<PostsContent onPostSelect={mockOnPostSelect} />);
    });

    expect(showLoading).toHaveBeenCalled();

    await waitFor(() => {
      const recentPost = screen.getByText('Second Post');
      const olderPost = screen.getByText('First Post');

      expect(recentPost).toBeInTheDocument();
      expect(olderPost).toBeInTheDocument();

      const postsOrder = [recentPost, olderPost];
      expect(postsOrder[0]).toBe(recentPost);
      expect(postsOrder[1]).toBe(olderPost);
    });
  });

  it('should match snapshot', async () => {
    await act(async () => {
      const { asFragment } = render(
        <PostsContent onPostSelect={mockOnPostSelect} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
