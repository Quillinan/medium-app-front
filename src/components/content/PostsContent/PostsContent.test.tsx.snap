// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostsContent > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-center h-full mt-4"
    data-testid="posts-content"
  >
    <div
      class="space-y-8"
    >
      <p>
        Nenhum post encontrado.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
